### DISCUSSION TITLE
DefaultValues with FieldArray

### DISCUSSION THREAD

**Discussion by orixjintaokoong** [react-hook-form/react-hook-form#11580] [Q&A]
The problem I'm solving
-
A CRUD application with prefilled forms after getting from an API from a promise, users can remove and add items into the array dynamically and update the item via an API on submit.

Current approach
-
After getting a response, I will render a form and set the values as the `defaultValues` in the option for `useForm`.

The issue
-
After setting as default, when I remove one of the existing item from the array and adding a new one, the newly added item which I expect to have all empty values, now are prefilled with the previously deleted item's values.

Workaround
-
To use some sort of onMount logic to set the values of the form using `form.setValue("fields", data.fields")`. However, this can get complex as the form grows.

Codesandbox (reproduce)
-
[codesandbox](https://codesandbox.io/p/sandbox/distracted-estrela-n45myx?file=%2Fsrc%2FApp.tsx%3A6%2C26&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cltcoa2tn00062a69ukmgq274%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cltcoa2tm00022a697p5lmfmo%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cltcoa2tm00032a691qq360jl%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cltcoa2tm00052a69y11xbk69%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cltcoa2tm00022a697p5lmfmo%2522%253A%257B%2522id%2522%253A%2522cltcoa2tm00022a697p5lmfmo%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cltcofb5p00022a69h3z5ikq5%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A6%252C%2522startColumn%2522%253A26%252C%2522endLineNumber%2522%253A6%252C%2522endColumn%2522%253A26%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cltcofb5p00022a69h3z5ikq5%2522%257D%252C%2522cltcoa2tm00052a69y11xbk69%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cltcoa2tm00042a69c6csgn8t%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522cltcoa2tm00052a69y11xbk69%2522%252C%2522activeTabId%2522%253A%2522cltcoa2tm00042a69c6csgn8t%2522%257D%252C%2522cltcoa2tm00032a691qq360jl%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522cltcoa2tm00032a691qq360jl%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Comment by orixjintaokoong** [comment]
My suspicion is the `defaultValues` are literally like the ones in native HTML elements, which is predefined if the field has `value == null`. Hence, when I append an empty object all the previous values are populated instead of using empty values. I'm just wondering if there is an alternative or a better way around the issue.

**Comment by Moshyfawn** [comment]
If you are appending a new array item, you should provide default values for each field in the object, as opposed to an empty object.

From [`useFieldArray` docs](https://react-hook-form.com/docs/usefieldarray):

> Important: append data is required and not partial.

<br class="Apple-interchange-newline">

```diff
- fields.append({})
+ fields.append({
+   name: "",
+   age: ""
+ })
```
